### ISSUE TITLE
How to creat many backbones in an specific architecture?

### ISSUE THREAD

**Issue by ZeHuiGong** [PaddlePaddle/PaddleDetection#976]
Some architectures need to initialize two or more backbones (e.g. ResNet50) when extracting features from the input images, and these backbones DO NOT share weights but can load pre-trained weights (such as pre-trained on ImageNet). Because in the PaddleDetection project, the backbone or neck passed into the architecture, e.g., Faster R-CNN is a python class object. In this case, how do I create two or more backbones that do not share weights, in the architecture?

**Comment by qingqing01** [comment]
@ZeHuiGong   Maybe can set the `weight_prefix_name` in backbone,  for example, set `weight_prefix_name` in ResNet https://github.com/PaddlePaddle/PaddleDetection/blob/release/0.3/ppdet/modeling/backbones/resnet.py#L71 .   But note,  if so, the ImageNet-pretrained will not load successfully,  since the difference weights name.  And, you want to write a new architecture ?

**Comment by ZeHuiGong** [comment]
@qingqing01 Thanks for your reply. Yes, I want to implement DetectoRS (https://arxiv.org/abs/2006.02334) in the PaddleDetection project. I have listened to your advice, and set 'weight_prefix_name' correspondingly when I want to use a different backbone. However, the loss seems stuck by a certain value (e.g. 4.3) for a few thousand iterations. I believe that this may be caused by the "gap" between these backbones. Because I do not change 'weight_prefix_name' in the first backbone, while setting it different for the second (not loading pre-trained weights since the different weights name). So, if I want to load pre-trained weights for the second backbone, I have to change the 'load_params' function? https://github.com/PaddlePaddle/PaddleDetection/blob/e47e826ed1e4cefcbe4b6bdb1f976c68faa91799/tools/train.py#L197

**Comment by qingqing01** [comment]
@ZeHuiGong  Refer to the DetectorRS implementation https://github.com/joe-siyuan-qiao/DetectoRS/blob/master/configs/DetectoRS/DetectoRS_mstrain_400_1200_r50_40e.py . It seems, only one backbones. 

For how to load two or more backbones' pre-trained weights,  if there are saved into one Program/Graph,  load the pre-trained directtly.   If not,  maybe need to merge them two one.  I'm not sure  I totaly get what you mean.  Could you paste a config?

**Comment by ZeHuiGong** [comment]
@qingqing01 You can see line 5 in the DetectoRS config file. (https://github.com/joe-siyuan-qiao/DetectoRS/blob/0fd2e967f8c9c169f6bd97b3ef67916a228c63ae/configs/DetectoRS/DetectoRS_mstrain_400_1200_r50_40e.py#L5). This parameter decides how many backbones (or time steps) you want to use. Or you can see the paper (https://arxiv.org/abs/2006.02334) to get detailed information.  Below is the code that I implement the DetectoRS. The main part lies in the 'extract_feat' function. And also, I have changed the code in 'resnet.py'. Would you please kindly take a look?
`
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from paddle import fluid
from paddle.fluid.param_attr import ParamAttr
from paddle.fluid.initializer import Xavier, Constant

from ppdet.core.workspace import register
from ppdet.utils.check import check_version
from .centernet import CenterNet


__all__ = ['RFPCenterNet']


class ASPP(object):

    def __init__(self, out_channels):
        super(ASPP, self).__init__()
        self.out_channels = out_channels
        self.kernel_sizes = [1, 3, 3, 1]
        self.dilations = [1, 3, 6, 1]
        self.paddinds = [0, 3, 6, 0]
        self.aspp_num = len(self.kernel_sizes)

    def __call__(self, x, name=''):
        avg_x = fluid.layers.adaptive_pool2d(x, 1, pool_type="avg", name=name + 'aspp_ada_gap')
        outs = []
        for idx in range(self.aspp_num):
            inp = avg_x if (idx == self.aspp_num - 1) else x
            out = fluid.layers.conv2d(
                inp,
                self.out_channels,
                filter_size=self.kernel_sizes[idx],
                stride=1,
                padding=self.paddinds[idx],
                dilation=self.dilations[idx],
                param_attr=ParamAttr(name=name + 'aspp_conv{}.w'.format(idx)),
                bias_attr=ParamAttr(initializer=Constant(0),
                                    name=name + 'aspp_conv{}.b'.format(idx)),
                act='relu',
                name=name + 'aspp_conv{}'.format(idx))
            outs.append(out)
        outs[-1] = fluid.layers.expand(outs[-1], [1, 1, outs[0].shape[2], outs[0].shape[3]])
        out = fluid.layers.concat(outs, axis=1)

        return out


@register
class RFPCenterNet(CenterNet):
    """Recursive feature pyramid. details in https://arxiv.org/abs/2006.02334.
    Args:
            single_scale (bool): a flag that represents whether use single scale feature (e.g., level3)
            or multi-scale feature fusion (fuse features across various resolutions) to predict
            the final heatmap and size.
    """

    def __init__(self,
                 backbone,
                 neck=None,
                 head='CenterHead',
                 rfp_steps=2,
                 fpn_levels=(),
                 stage_with_rfp=[3, 4, 5],
                 rfp_sharing=False,
                 num_classes=80,
                 single_scale=True):
        check_version('1.8.0')
        assert neck is not None, 'in recursive feature pyramid, you must have a fpn neck!'
        super(RFPCenterNet, self).__init__(
            backbone,
            neck,
            head,
            num_classes,
            single_scale)

        fpnidx_list = list(range(fpn_levels[1], fpn_levels[0]-1, -1))
        self.stage2fpnidx = {stage: fpnidx_list.index(stage) for stage in stage_with_rfp}
        self.rfp_steps = rfp_steps
        self.rfp_sharing = rfp_sharing
        self.stage_with_rfp = stage_with_rfp
        self.rfp_aspp = ASPP(neck.num_chan // 4)

    def extract_feat(self, x):
        input_w = x.shape[-1]
        # step 1
        self.backbone.prefix_name = ''
        body_feats = self.backbone(x)
        body_feats, _ = self.neck.get_output(body_feats)
        body_feats = tuple(body_feats.values())

        # feedback connection (recursive refinement)
        for rfp_idx in range(self.rfp_steps - 1):
            rfp_feats = tuple(self.rfp_aspp(body_feats[self.stage2fpnidx[i]])
                              for i in range(2, 6) if i in self.stage_with_rfp)
            if self.rfp_sharing:  # sharing backbone parameters
                body_feats_idx = self.backbone(x, rfp_feats)
            else:
                self.backbone.prefix_name = 'rfp_step{}'.format(rfp_idx + 1)
                body_feats_idx = self.backbone(x, rfp_feats)
            body_feats_idx, _ = self.neck.get_output(body_feats_idx)
            body_feats_idx = tuple(body_feats_idx.values())
            body_feats_new = []
            # feature fusion, fuse features from
            for ft_idx in range(len(body_feats_idx)):
                add_weight = self.rfp_weight(body_feats_idx[ft_idx])
                body_feats_new.append(add_weight * body_feats_idx[ft_idx] +
                                      (1-add_weight) * body_feats[ft_idx])
            body_feats = body_feats_new
        body_feats = self.fuse_features(body_feats, input_w)

        return body_feats

    def rfp_weight(self, feat, name=''):
        add_weight = fluid.layers.conv2d(
            feat,
            1,
            filter_size=1,
            stride=1,
            padding=0,
            param_attr=ParamAttr(
                initializer=Constant(0),
                name=name + 'rfp_weight.w'),
            bias_attr=ParamAttr(
                initializer=Constant(0),
                name=name + 'rfp_weight.b'),
            name=name + 'rfp_weight')
        add_weight = fluid.layers.sigmoid(add_weight)

        return add_weight

`

**Comment by qingqing01** [comment]
@ZeHuiGong  From the source code

https://github.com/joe-siyuan-qiao/DetectoRS/blob/0fd2e967f8c9c169f6bd97b3ef67916a228c63ae/mmdet/models/detectors/rfp.py#L82-L84

the multiple backbones  share the same weights,  They load the same pre-trained weights,     pretrained='torchvision://resnet50' .  I think there is no need to set `weight_prefix_name`.

**Comment by WuShuang1998** [comment]
Maybe you can set extra backbones in config file 
 ```python
RFP:
  rfp_steps: 2
  min_level: 2
  max_level: 6
  num_chan: 256
  spatial_scale: [0.03125, 0.0625, 0.125, 0.25]
  rfp_modules: DetectoRS_ResNet_rfp

DetectoRS_ResNet_rfp:
  weight_prefix_name: rfp_
  norm_type: affine_channel
  norm_decay: 0.
  depth: 50
  feature_maps: [2, 3, 4, 5]
  freeze_at: 2
  output_img: True
 ```

then modify the function load_params() in ppdet/utils/checkpoint.py to load pre-trained weight correctly
 ```python
def load_params(exe, prog, path, ignore_params=[]):
    """
    Load model from the given path.
    Args:
        exe (fluid.Executor): The fluid.Executor object.
        prog (fluid.Program): load weight to which Program object.
        path (string): URL string or loca model path.
        ignore_params (list): ignore variable to load when finetuning.
            It can be specified by finetune_exclude_pretrained_params 
            and the usage can refer to docs/advanced_tutorials/TRANSFER_LEARNING.md
    """

    if is_url(path):
        path = _get_weight_path(path)

    path = _strip_postfix(path)
    if not (os.path.isdir(path) or os.path.exists(path + '.pdparams')):
        raise ValueError("Model pretrain path {} does not "
                         "exists.".format(path))

    logger.debug('Loading parameters from {}...'.format(path))

    ignore_set = set()
    state = _load_state(path)

    state_add = {}
    for k, v in state.items():
        if 'bn' in k or 'res' in k:
            state_add['rfp_' + k] = v
    state.update(state_add)
    del state_add

    # ignore the parameter which mismatch the shape 
    # between the model and pretrain weight.
    all_var_shape = {}
    for block in prog.blocks:
        for param in block.all_parameters():
            all_var_shape[param.name] = param.shape
    # print(all_var_shape)
    ignore_set.update([
        name for name, shape in all_var_shape.items()
        if name in state and shape != state[name].shape
    ])

    if ignore_params:
        all_var_names = [var.name for var in prog.list_vars()]
        ignore_list = filter(
            lambda var: any([re.match(name, var) for name in ignore_params]),
            all_var_names)
        ignore_set.update(list(ignore_list))

    if len(ignore_set) > 0:
        for k in ignore_set:
            if k in state:
                logger.warning('variable {} not used'.format(k))
                del state[k]
    fluid.io.set_program_state(prog, state)

**Comment by ZeHuiGong** [comment]
Thanks for your kindly reply, I wonder where is the source code corresponding to the config file listed below"
'''
RFP:
 rfp_steps: 2
 min_level: 2
 max_level: 6
 num_chan: 256
 spatial_scale: [0.03125, 0.0625, 0.125, 0.25]
 rfp_modules: DetectoRS_ResNet_rfp

DetectoRS_ResNet_rfp:
 weight_prefix_name: rfp_
 norm_type: affine_channel
 norm_decay: 0.
 depth: 50
 feature_maps: [2, 3, 4, 5]
 freeze_at: 2
 output_img: True
'''
Would you kindly please share the source code link?

**Comment by ZeHuiGong** [comment]
@WuShuang1998 
Thanks for your kindly reply, I wonder where is the source code corresponding to the config file listed below"
`
RFP:
rfp_steps: 2
min_level: 2
max_level: 6
num_chan: 256
spatial_scale: [0.03125, 0.0625, 0.125, 0.25]
rfp_modules: DetectoRS_ResNet_rfp

DetectoRS_ResNet_rfp:
weight_prefix_name: rfp_
norm_type: affine_channel
norm_decay: 0.
depth: 50
feature_maps: [2, 3, 4, 5]
freeze_at: 2
output_img: True
`
Would you kindly please share the source code link?

**Comment by WuShuang1998** [comment]
Hi @ZeHuiGong,
I'm implementing RFP in the PaddleDetection framework, but the result is not good. We can discuss it by email if you like.

**Comment by ZeHuiGong** [comment]
@WuShuang1998 Me too!! The reimplementation results are far from the original paper. As for me, the biggest problem is the training loss does not drop for a long time. It seems stuck by a certain value, and the results are not satisfactory. Because I am newer for the PaddlePaddle framework, maybe there are some details that we ignore? If you like , my email address is at: zehuigong@foxmail.com.
